### DISCUSSION TITLE
CLI sleap-track -  save as you go option

### DISCUSSION THREAD

**Discussion by WeissShahaf** [talmolab/sleap#1585] [Ideas]
Dear Devs,
i'm trying to predict on very long videos. and i'm running into issues:
1. i can't find the videos being inferred. i'm guessing it's only writing the file at the end. which for big files may be problematic.
- is it possible to implement an optional feature, similiar to "use_shelve" on "deeplabcut.analyze_videos", where: 
"data are written to disk on the fly using a “shelf”; i.e., a pickle-based, persistent, database-like object by default, resulting in constant memory footprint."

2.  using --format analsis.csv still ends up with an .h5 file

thank you,
Shahaf

**Comment by talmo** [comment]
Hi @WeissShahaf,

Apologies for the delay!

Yes, this could be a great idea for when you have very long videos or limits on your memory.

We don't quite have something like this right now and it'll probably be a bit tricky given our current data model and HDF5-based data format, but we'd welcome contributions.

My recommendation would be to use the low-level inference API and handle the online serialization (saving) yourself using your preferred method. [This notebook is a good starting point](https://sleap.ai/notebooks/Interactive_and_realtime_inference.html#interactive-inference).

It might look something like this:

```py
import sleap
import shelve

video = sleap.load_video("video.mp4")
predictor = sleap.load_model("models/my_model")

n_frames = len(video)
batch_size = 4

writer = shelve.open("predictions")
for i0 in range(0, n_frames, batch_size):
    i1 = min(i0 + batch_size, n_frames)
    frames = video[i0:i1]
    writer[f"frames{i0}-{i1}"] = predictor.inference_model.predict(frames)
writer.close()
```

Feel free to let us know if you have any questions!!

Cheers,

Talmo
